### ISSUE TITLE
Enhancements in the UiUtil.ShowSubtitle() method

### ISSUE THREAD

**Issue by MohammadHamdyGhanem** [SubtitleEdit/subtitleedit#3353]
Hi @niksedk:
I made some enhancements in the UiUtil.ShowSubtitle() method, because it makes the app consumes a lot of cpu power even if the video is paused!

1- I saved the last paragraph index and the last video Position, so:
  if the new pos is next to the last pos, we can start searching for the subtitle from lastParagraphIndex to the end of the list
   otherwise, we search from the beginning to lastParagraphIndex

2- if we reach a paragraph that starts after the video pos, then no need to complete the loop: break.

The above two enhancements, saves a lot of time, which can be huge in long subtitle files (long verbose videos).
Another enhancement is increasing the ShowSubtitleTimer.Interval from 16ms to 50ms. No human eye can notice the difference.

3- To solve the problem of very short duration subtitles, I ensured to view each subtitle for a MinDuration (I choose to be 1.5 second) if possible. If this causes an overlapping with the next paragraph, I trim the MinDuration so end before the next paragraph. You can add an option to the options form to allow the user to control the MinDuration period.
Note that In my VASE I allow the user to change the play rate of the video, so I scale the MinDuration to be increased with higher play rates. If you didn't have this in SE, just make rate = 1;
  
```
        static int lastParagraphIndex = 0;
        static double lastPositionMs = 0;
        static string lastText = "";
        public static int ShowSubtitle(Subtitle subtitle, VideoPlayerContainer videoPlayerContainer)
        {
            const int MinDuration = 1500;
            if (videoPlayerContainer.VideoPlayer != null)
            {
                var rate = 1; //videoPlayerContainer.PlayRate;
                if (rate < 1)
                    rate = 1;

                var pCount = subtitle.Paragraphs.Count;
                double positionMs = (videoPlayerContainer.CurrentPosition * TimeCode.BaseUnit) + 5;
                
                var startIndex = 0;
                var endIndex = lastParagraphIndex;

                if(positionMs >= lastPositionMs)
                {
                    startIndex = lastParagraphIndex;
                    endIndex = pCount;
                }
                lastPositionMs = positionMs;

                for (int i = startIndex; i < endIndex; i++)
                {
                    var p = subtitle.Paragraphs[i];
                    var nextStart = (i < pCount - 1) ? subtitle.Paragraphs[i + 1].StartTime.TotalMilliseconds : p.EndTime.TotalMilliseconds +1;
                    var pStart = p.StartTime.TotalMilliseconds;
                    var d = p.Duration.TotalMilliseconds * rate;
                    if (d < MinDuration * rate)
                        d = MinDuration * rate;
                    var pEnd = Math.Min(pStart +d, nextStart - 1);

                    if (pStart <= positionMs)
                    {
                        if (pEnd > positionMs)
                        {
                            string text = p.Text.Replace("|", Environment.NewLine);
                            bool isInfo = p == subtitle.Paragraphs[0] && (p.StartTime.TotalMilliseconds == 0 && p.Duration.TotalMilliseconds == 0 || p.StartTime.TotalMilliseconds == Pac.PacNullTime.TotalMilliseconds);
                            if (!isInfo)
                            {
                                if (videoPlayerContainer.LastParagraph != p)
                                    videoPlayerContainer.SetSubtitleText(text, p, subtitle);
                                else if (lastText != text)
                                    videoPlayerContainer.SetSubtitleText(text, p, subtitle);

                                lastParagraphIndex = i;
                                lastText = text;
                                return i;
                            }

                            lastText = text;
                        }
                    }
                    else
                    {
                        lastParagraphIndex = Math.Max(i - 1, 0);
                        break;
                    }
                }

                if (!string.IsNullOrEmpty(videoPlayerContainer.SubtitleText))
                    videoPlayerContainer.SetSubtitleText(string.Empty, null, subtitle);
            }
            return -1;
        }
```

**Comment by marb99** [comment]
@MohammadHamdyGhanem 

These changes cause multiple errors when compiling.

![1](https://user-images.githubusercontent.com/7856897/52528627-e64fe880-2cda-11e9-8e55-967d2c33fcfc.PNG)

**Comment by MohammadHamdyGhanem** [comment]
Sorry. I added some shortcut properties to the paragraph class because I use them a lot. You can add them, or just replace:
- `p.StartTimeMs` with `p.StartTime.TotalMilliseconds`
- `p.DurationMs` with 'p.Duration.TotalMilliseconds`
- 'p.EndTimeMs' with p.EndTime.TotalMilliseconds'.

**Comment by niksedk** [comment]
@MohammadHamdyGhanem: thx for the input - how is the cpu usage on your computer with video paused? On my computer (5-6 years old) the SE cpu usage is about 3-5%.

I don't think that saving last paragraph index will benefit much, as looping though10.000 paragraphs + check for each start/end takes less than one millisecond.

Changing ShowSubtitleTimer.Interval from 17ms to 50ms or higher will use less cpu and would probably help if you're running on a very slow pc.

**Comment by MohammadHamdyGhanem** [comment]
@niksedk:
It is always better to write optimized code.. This function with your clock-timing is called 62 times per second, executing a loop and formatting rtf text each time. I lived with your code for more than one year on my old PC, but I reviewed it to adjust the duration with higher play rates, and added the min duration part, this is when I noticed that the loop can be reduced to a one iteration most of the time (normal video play without jumping by the progress bar) just by saving the last index. This is some sort of binary search.
One Benefit of this optimization is making tracing and debugging easier. Many times I face a problem and need to see what happens, but I stuck in these endless loops! 
In fact, there is one more optimization: Stop the ShowSubtitleTimer when the video is paused, put this will need reviewing all places where user can edit the subtitle text , color, font and etc to refresh the displayed subtitle!.. I am not aware of all of these places, so I didn't want to take risk!
This should apply to all timers (and there are many!): Stop any timer when it is not wanted. 
You did a great work and I'm thankful.
Thanks.
